### DISCUSSION TITLE
Missing or invalid "eth_signTypedData_v4"

### DISCUSSION THREAD

**Discussion by trevormil** [wevm/wagmi#2240] [Q&A]
I am using Web3Modal and wagmi. When I call signTypedData() on desktop with Metamask, it works fine (same with Metamask mobile). When I attempt to call it with Rainbow wallet and Zerion wallet on mobile, I get this error: "Missing or invalid. request() method: eth_signTypedData_v4”

Am I doing something wrong or is this unsupported?

Here is my _app.tsx.

```
import { EthereumClient, w3mConnectors, w3mProvider } from '@web3modal/ethereum'
import { Web3Modal } from '@web3modal/react'
import { configureChains, createClient, WagmiConfig } from 'wagmi'
import { mainnet } from 'wagmi/chains'

const chains = [mainnet]
const projectId = 'febf8d9986a2cd637fa4004338dad39b'

const { provider } = configureChains(chains, [w3mProvider({ projectId })])
const wagmiClient = createClient({
    autoConnect: true,
    connectors: w3mConnectors({ projectId, version: 1, chains }),
    provider,
})
const ethereumClient = new EthereumClient(wagmiClient, chains)
const queryClient = new QueryClient()
```

and here is my sign function:

```
import { disconnect as disconnectWeb3, signMessage, signTypedData } from "@wagmi/core";
...
...
const sig = await signTypedData({
    value: txn.eipToSign.message,
    types: txn.eipToSign.types,
    domain: txn.eipToSign.domain
});
```

Seems like they have the same problem: [https://forum.moralis.io/t/how-call-signtypeddata-when-i-use-web3modal/22927](https://forum.moralis.io/t/how-call-signtypeddata-when-i-use-web3modal/22927).

**Comment by trevormil** [comment]
I should also note that signMessage() works fine with both Rainbow Wallet and Zerion Wallet. It is only the signTypedData.

**Comment by cosiato** [comment]
Same issue here, getting this message while using useSignTypedData with Zengo and Trustwallet apps.

Wagmi version 1.0.9 and WalletConnect v2.

**Reply by TimDaub** [reply]
I'm getting this also with wagmi 0.12.13 and connectkit 1.3.0, attempting to upgrade walletconnect 2.0

**Reply by TimDaub** [reply]
btw. the problem is fixed with connect kit if you intentionally NOT use walletconnect 2.0
```
const client = createClient(
  getDefaultClient({
    appName: "Kiwi News",
    alchemyId: "hidden",
    // NOTE: This is intentionally commented
    //walletConnectProjectId: "hidden",
  })
);
```

**Reply by TimDaub** [reply]
@pedroid ☝🏼

**Reply by cosiato** [reply]
I investigated this issue a bit more and it seems WalletConnect v2 does not support RPC function `eth_signTypedData_v4`, only `eth_signTypedData`.  Info [here ](https://docs.walletconnect.com/1.0/json-rpc-api-methods/ethereum).

I suspect they are blocking the eth_signTypedData_v4 on their side.

**Reply by TimDaub** [reply]
how to fix for wc2?

**Reply by pedrouid** [reply]
Hey @tomtomdu73 that RPC reference does not affect our operation that is just informational and you can see that we provide a list of RPC methods that can be used as `optionalMethods` wiht our ethereum-provider: https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/providers/ethereum-provider/src/constants/rpc.ts#L2

I'm still confused why wagmi is causing any issues here since they include all methods listed in that file by default

**Reply by TimDaub** [reply]
what alphawallet support wrote me yesterday

![IMG_2816](https://github.com/wagmi-dev/wagmi/assets/2758453/f4c7313f-90a3-4397-a8c7-2d7f7675446e)

**Reply by TimDaub** [reply]
Trustwallet now works for me on iOS

**Comment by Jabher** [comment]
Any news on that?

**Reply by TimDaub** [reply]
my observation has been that a few wallets are still buggy. My app is using EIP712 signing a lot and most wallets I tested today worked OK. But AlphaWallet was definitely broken for me on Android and iOS

**Comment by GarrettJMU** [comment]
Any update on this? Feels like I can't be only one experiencing this pain.

**Comment by web-crab** [comment]
Fix for WalletConnect v2 & Trust Wallet. It works for me

```ts
import { EthereumProvider } from '@walletconnect/ethereum-provider';
import { WalletConnectConnector } from '@wagmi/core/connectors/walletConnect';

let wcProvider: any;

WalletConnectConnector.prototype.getProvider = async function ({ chainId } = {}) {
  if (!wcProvider) {
    const [defaultChain, ...optionalChains] = this.chains.map((chain) => chain.id);

    wcProvider = await EthereumProvider.init({
      projectId,
      chains: [defaultChain],
      optionalChains: optionalChains,
      showQrModal: false,
      methods: ['eth_sendTransaction', 'personal_sign', 'eth_signTypedData_v4'],
      rpcMap: Object.fromEntries(
        this.chains.map((chain) => [chain.id, chain.rpcUrls.default.http[0]]),
      ),
    });
  }

  if (chainId) {
    await this.switchChain(chainId);
  }

  return wcProvider;
};
```

**Reply by abdtimur** [reply]
Worked for me as well. Thanks! 
I tried to change `optionalMethods` in `EthereumProviderOptions` first, but it didn't help. Only manual `methods` overwrite.
Looks like if WalletConnectConnector could set it by default or allow to pass `methods` as param, it would solve the issue.

**Reply by GuiDupont** [reply]
on my side this solution breaks wagmi capacity to identify network changes.

**Reply by web-crab** [reply]
me too. check updated comment

**Comment by paintoshi** [comment]
The solution that worked for me with using the web3modal, was to add Ethereum MainNet to the config. I was only using Fantom as the configured network in the Web3Modal. I don't need mainnet but I can't see how it would do any harm sitting in there. So basically just:

```
import { fantom, mainnet } from 'wagmi/chains'
additionalChains = [fantom, mainnet]
const { publicClient } = configureChains(additionalChains, [w3mProvider({ projectId })])
const wagmiConfig = createConfig({
  autoConnect: true,
  connectors: w3mConnectors({ projectId, chains: additionalChains }),
  publicClient,
})

const ethereumClient = new EthereumClient(wagmiConfig, additionalChains)
```

Using:

```
"@web3modal/ethereum": "^2.7.1",
"@web3modal/react": "^2.7.1",
"viem": "^1.4.2",
"wagmi": "^1.3.9",
```

Now the signature works over WalletConnect (using Fantom).

```
signTypedDataAsync({
  domain,
  primaryType: 'xxx',
  types,
  message: message as unknown as { [key: string]: unknown },
})
  .then((signature) => {
    console.info(signature)
  })
  .catch((err) => {
    console.error(err)
  })
```

**Reply by zation** [reply]
I guess the key point is not `mainnet`. There must be more or equal than 2 chains in the config to make `optionalChains` is not `undefined` when init `EthereumProvider `. This problem is related to WalletConnect namespaces.
Refer to https://github.com/wagmi-dev/wagmi/issues/3130 and https://github.com/wagmi-dev/wagmi/issues/3012
